core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -105,6 +107,12 @@ public MessageFramer setCompressor(Compressor compressor) {
     return this;
   }
 
+  @Override
+  public Framer setStreamCompression(boolean enable) {
+    streamCompression = enable;
```

Comment:
The stream compression code is tighly integrated and seems error-prone. Would it make more sense to make a "fake" Sink that does the gzipping and then delegates to the real one? We'd also need to fake the WritableBuffer to be byte[]-based.

That would, however, be pretty inefficient and require a lot more copying. If we created a new sink-like API (let's call it "better sink") but that also included a `write(byte[], int, int)`, then it seems we could clean up the code while being efficient. Basically we'd move `buffer`/writeRaw/commitToSink (anything WritableBuffer-related) into the "default" better-sink, so that the code acts more like a stream.

Reply:
Because it is non-obvious, I don't think we'd include anything concerning `messageCompression` in better-sink. Instead, when `messageCompression` changes, we'd call the gzip sink directly (knowing its class, not the interface) to inform it of the new setting. Although that could be included in the interface, it probably doesn't gain much by doing so.